pkg/networkservice/connectioncontext/dnscontext/vl3dns/server.go

```diff
@@ -128,19 +128,18 @@ func (n *vl3DNSServer) Request(ctx context.Context, request *networkservice.Netw
 		}
 	}
 
-	ips := getSrcIPs(request.GetConnection())
-	if len(ips) > 0 {
-		for _, recordName := range recordNames {
-			n.dnsServerRecords.Store(recordName, ips)
-		}
-
-		metadata.Map(ctx, false).Store(clientDNSNameKey{}, recordNames)
-	}
-
 	resp, err := next.Server(ctx).Request(ctx, request)
 	if err == nil {
 		configs := make([]*networkservice.DNSConfig, 0)
 		if srcRoutes := resp.GetContext().GetIpContext().GetSrcRoutes(); len(srcRoutes) > 0 {
+			ips := getSrcIPs(resp)
+			if len(ips) > 0 {
+				for _, recordName := range recordNames {
+					n.dnsServerRecords.Store(recordName, ips)
+				}
+
+				metadata.Map(ctx, false).Store(clientDNSNameKey{}, recordNames)
+			}
```

Comment:
I think it would be better to paste this block after line 132 (`if err == nil {`)